sw/host/tests/usbdev/usbdev_stream/usbdev_int.cc

```diff
@@ -0,0 +1,275 @@
+    std::cerr << "" << std::endl;
+  }
+}
+
+bool USBDevInt::Resume() {
+  SetClosing(false);
+
+  int rc = dev_->ClaimInterface(interface_);
+  if (rc < 0) {
+    return dev_->ErrorUSB("ERROR: Claiming interface", rc);
+  }
+  return true;
+}
+
+// Return a summary report of the stream settings of status.
+std::string USBDevInt::Report(bool status, bool verbose) const { return ""; }
+
+void USBDevInt::DumpIntTransfer(struct libusb_transfer *xfr) const {
+  const void *buf = reinterpret_cast<void *>(xfr->buffer);
+  std::cout << "Buffer " << buf << " length " << xfr->length
+            << " => actual length " << xfr->actual_length << std::endl;
+  buffer_dump(stdout, xfr->buffer, xfr->actual_length);
+}
+
+// Retrieving of IN traffic from device.
+bool USBDevInt::ServiceIN() {
+  // Ensure that we have enough space available for a full packet; the device
+  // software decides upon the length of each packet.
+  uint8_t *space;
+  bool ok = ProvisionSpace(&space, maxPacketSize_);
+  if (ok) {
+    uint32_t to_fetch = maxPacketSize_;
+    if (to_fetch > transfer_bytes_ - bytes_recvd_) {
+      to_fetch = transfer_bytes_ - bytes_recvd_;
+    }
+
+    if (!xfrIn_) {
+      xfrIn_ = dev_->AllocTransfer(0U);
+      if (!xfrIn_) {
+        return false;
+      }
+    }
+
+    if (bulk_) {
+      dev_->FillBulkTransfer(xfrIn_, epIn_, space, to_fetch, CbStubIN, this,
+                             kDataTimeout);
+    } else {
+      dev_->FillIntTransfer(xfrIn_, epIn_, space, to_fetch, CbStubIN, this,
+                            kDataTimeout);
+    }
+
+    int rc = dev_->SubmitTransfer(xfrIn_);
+    if (rc < 0) {
+      return dev_->ErrorUSB("ERROR: Submitting IN transfer", rc);
+    }
+    inActive_ = true;
+  } else {
+    inActive_ = false;
+  }
+
+  return true;
+}
+
+// Sending of OUT traffic to device.
+bool USBDevInt::ServiceOUT() {
+  // Do we have any data ready to send?
+  uint8_t *data;
+  uint32_t num_bytes = DataAvailable(&data);
+  if (num_bytes > 0U) {
+    // Supply details of the single OUT packet.
+    if (!xfrOut_) {
+      xfrOut_ = dev_->AllocTransfer(0U);
+      if (!xfrOut_) {
+        // Stream is not operational.
+        return false;
+      }
+    }
+
+    if (bulk_) {
+      dev_->FillBulkTransfer(xfrOut_, epOut_, data, num_bytes, CbStubOUT, this,
+                             kDataTimeout);
+    } else {
+      dev_->FillIntTransfer(xfrOut_, epOut_, data, num_bytes, CbStubOUT, this,
+                            kDataTimeout);
+    }
+
+    int rc = dev_->SubmitTransfer(xfrOut_);
+    if (rc < 0) {
+      return dev_->ErrorUSB("ERROR: Submitting OUT transfer", rc);
+    }
+    outActive_ = true;
+  } else {
+    // Nothing to propagate at this time.
+    outActive_ = false;
+  }
+  // Stream remains operational, even if it presently has no work on the OUT
+  // side.
+  return true;
+}
+
+bool USBDevInt::Service() {
+  if (failed_) {
+    return false;
+  }
+  // (Re)start Isochronous IN traffic if not already in progress.
```

Comment:
I think the comment is wrong because this class handles interrupt and bulk tranfers?